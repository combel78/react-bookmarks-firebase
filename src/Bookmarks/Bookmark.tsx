import React from "react";

import { makeStyles } from '@material-ui/core/styles';
import EditIcon from '@material-ui/icons/Edit';
import { Grid, Paper, Typography, Link, IconButton } from "@material-ui/core";
import BookmarkEditForm from "./BookmarkEditForm";
import { useUserContext } from './BookmarkHome';

const useStyles = makeStyles(theme => ({
    paper: {
        padding: theme.spacing(2),
        textAlign: 'center',
        color: theme.palette.text.secondary,
    }
}));

interface BookmarkProps {
    bookmark: BookmarkType,
    onBookmarkSave: any,
    onBookmarkDelete: any,
}

const Bookmark: React.FC<BookmarkProps> = (bookmarkProps) => {

    const classes = useStyles();

    const user = useUserContext();

    const [currBookmark, setCurrentBookmark] = React.useState(bookmarkProps.bookmark);

    const [editFormOpen, setEditFormOpen] = React.useState(false);

    const handleBookmarkEditFormOpen = () => {
        setEditFormOpen(true);
    }

    const handleBookmarkEditFormClose = () => {
        setEditFormOpen(false);
    }

    const handleBookmarkEditFormSave = (bookmark: BookmarkType) => {
        setCurrentBookmark(bookmark);
        setEditFormOpen(false);
        bookmarkProps.onBookmarkSave(bookmark);
    }

    const handleBookmarkEditFormDelete = (bookmark: BookmarkType) => {
        setEditFormOpen(false);
        bookmarkProps.onBookmarkDelete(bookmark);
    }

    return(
        <React.Fragment>
            <BookmarkEditForm open={editFormOpen} bookmark={currBookmark} onClose={handleBookmarkEditFormClose} onSave={handleBookmarkEditFormSave} onDelete={handleBookmarkEditFormDelete} />
            <Grid container spacing={6}>
                <Grid item xs={12}>
                    <Paper className={classes.paper}>
                        <Grid container>
                            <Grid item xs={10}>
                                <Typography variant="h6" align="left" gutterBottom>
                                    <Link data-testid="bookmark-link-element" href={currBookmark.url} target="_blank" rel="noreferrer">{currBookmark.name}</Link>
                                </Typography>
                            </Grid>
                            <Grid item xs={2}>
                                <Grid container justify="flex-end">
                                { user.isLoggedIn &&
                                    <IconButton size="small" color="primary" onClick={handleBookmarkEditFormOpen}>
                                        <EditIcon />
                                    </IconButton>
                                }
                                </Grid>
                            </Grid>
                            <Grid item xs={12}>
                                <Typography variant="subtitle1" align="left" data-testid="bookmark-descr-element">{currBookmark.description}</Typography>
                            </Grid>                            
                        </Grid>
                    </Paper>
                </Grid>
            </Grid>
        </React.Fragment>
    )
}

export default Bookmark;