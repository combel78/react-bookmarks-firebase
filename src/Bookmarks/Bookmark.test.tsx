import React from 'react';
import ReactDOM from 'react-dom';
import Bookmark from './Bookmark';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';

const bookmark: BookmarkType = {
    id: 1,
    name: 'React Homepage',
    url: 'https://reactjs.org',
    description: 'gehe zur React-Homepage'
}

const handleBookmarkSave = jest.fn();
const handleBookmarkDelete = jest.fn();

afterEach(cleanup);

it('Test 1: Bookmark rendert ohne Crash', () => {
    const divElement = document.createElement('divElement');
    ReactDOM.render(
        <Bookmark bookmark={bookmark} key={bookmark.id} onBookmarkSave={handleBookmarkSave} onBookmarkDelete={handleBookmarkDelete} />,
        divElement
    );
    ReactDOM.unmountComponentAtNode(divElement);
});

it('Test 2: prÃ¼fe Bookmark-Ausgabewerte', () => {
    const {getByTestId: bookmarkComponent} = render(<Bookmark bookmark={bookmark} key={bookmark.id} onBookmarkSave={handleBookmarkSave} onBookmarkDelete={handleBookmarkDelete} />);
    //lade UI-Elemente per data-testid-Attribut aus Bookmark.tsx
    //Funktionen toHaveTextContent und toHaveAttribute stammen aus @testing-library/jest-dom/extend-expect
    expect(bookmarkComponent('bookmark-link-element')).toHaveTextContent(bookmark.name);
    expect(bookmarkComponent('bookmark-link-element')).toHaveAttribute('href', bookmark.url);
    expect(bookmarkComponent('bookmark-descr-element')).toHaveTextContent(bookmark.description);
});

it('Test 3: Komponente entspricht Snapshot', () => {
    const bookmarkComponent = renderer.create(<Bookmark bookmark={bookmark} key={bookmark.id} onBookmarkSave={handleBookmarkSave} onBookmarkDelete={handleBookmarkDelete} />);
    expect(bookmarkComponent).toMatchSnapshot();
})