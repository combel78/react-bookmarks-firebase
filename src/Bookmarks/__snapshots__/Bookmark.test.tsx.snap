// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 3: Komponente entspricht Snapshot 1`] = `
Array [
  <div />,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-6"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiPaper-root makeStyles-paper-7 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
          >
            <h6
              className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom MuiTypography-alignLeft"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                data-testid="bookmark-link-element"
                href="https://reactjs.org"
                onBlur={[Function]}
                onFocus={[Function]}
                rel="noreferrer"
                target="_blank"
              >
                React Homepage
              </a>
            </h6>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          >
            <div
              className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-flex-end"
            />
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <h6
              className="MuiTypography-root MuiTypography-subtitle1 MuiTypography-alignLeft"
              data-testid="bookmark-descr-element"
            >
              gehe zur React-Homepage
            </h6>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
